File: Bot/Commands/CallbackQueryHandler.py
from datetime import datetime, timedelta
from typing import Optional

from telethon import TelegramClient
from telethon.errors.rpcerrorlist import PeerIdInvalidError
from telethon.tl import functions, types
from telethon.tl.functions.messages import GetFullChatRequest
from telethon.tl.types import InputPeerChannel
from telethon.types import UpdateNewChannelMessage

from apihelper.BotSettingsServiceApi import BotSettingsServiceApi
from apihelper.PostServiceApi import PostServiceApi
from Commands.MessageStatus import MessagesStatus
from utility.inline_replies import *


class CallbackQueryHandler:
    def __init__(self, client: TelegramClient,user_client: TelegramClient,
                 bot_settings_service : BotSettingsServiceApi, 
                 post_service : PostServiceApi):
        
        self.user_client = user_client
        self.client = client
        self.bot_settings_service = bot_settings_service
        self.post_service = post_service
        
        self.bot_settings = self.bot_settings_service.botsettings

    async def update_forwarded_post_status(self, post_info_entity, status: str):
        post_events = self.post_service.get_removable_events(post_info_entity["id"])

        for post_event in post_events:
            try:
                await self.client.edit_message(
                    post_event["telegramId_ChatId"],
                    message=post_event["telegramId_PostId"],
                    text=status,
                )
            except Exception as ex:
                print(f"Error editing message: {ex}")
                continue
        
    async def on_set_bridge_channel(self, message, bridge_channel_id: str):
        self.bot_settings = self.bot_settings_service.get_bot_settings()
        
        chat_id = message.chat_id
        message_id = message.message_id

        new_status = "<b>این کانال به عنوان کانال میانی تنظیم شد ✅</b>"

        self.bot_settings["bridgeChanneID"] = int(bridge_channel_id)

        # Call the service to set bot settings
        self.bot_settings = self.bot_settings_service.set_bot_settings(self.bot_settings)

        try:
            # Edit the message text
            await self.client.edit_message(
                chat_id,
                message=message_id,
                text=new_status,
                parse_mode='html'
            )
        except Exception as e:
            print(f"Error editing message: {e}")
            
            
    async def on_accept_post(self, message):
        self.bot_settings = self.bot_settings_service.get_bot_settings()
        
        data = message.data.decode('utf-8').split("/")
        post_info_entity = self.post_service.get_post_by_id(int(data[1]))

        accepted_post = self.post_service.accept_post(post_info_entity)

        chat_id = message.chat_id
        message_id = message.message_id
        first_name = message.sender.first_name
        username = message.sender.username or ""

        new_status = MessagesStatus.accept_message(first_name , username)

        try:
            date_format = "%m/%d/%Y %H:%M:%S"
            date_time = datetime.strptime(accepted_post["forwardToChannelDate"], date_format)
        except:
            date_format = '%Y-%m-%dT%H:%M:%S.%fZ'
            date_time = datetime.strptime(accepted_post["forwardToChannelDate"], date_format)
            
        fwd_msg_events = await self.user_client(functions.messages.ForwardMessagesRequest(
            from_peer=post_info_entity["telegramId_ChatId"],
            id=post_info_entity['telegram_mediagroup_postids'],
            to_peer=self.bot_settings["mainChanneID"],
            schedule_date=date_time
        ))
        
        for updt in fwd_msg_events.updates:
            if type(updt) == UpdateNewChannelMessage and updt.message != '':
                sent_message = updt.message
                break

        # log_message = f"Forward to channel: {post_info_entity.telegram_id_post_id}, {post_info_entity.telegram_id_chat_id}, postid: {post_info_entity.id}"
        # logger.log_info(log_message)

        await self.client.edit_message(
            chat_id,
            message=message_id,
            text=new_status
        )

        await self.update_forwarded_post_status(post_info_entity, new_status)
        
    async def on_decline_post(self, message):
        self.bot_settings = self.bot_settings_service.get_bot_settings()
        
        data = message.data.decode('utf-8').split("/")
        post_info_entity = self.post_service.get_post_by_id(int(data[1]))

        chat_id = message.chat_id
        message_id = message.message_id
        first_name = message.sender.first_name
        username = message.sender.username or ""

        self.post_service.decline_post(post_info_entity)

        new_status = MessagesStatus.decline_message(first_name, username)

        await self.client.edit_message(
            chat_id,
            message=message_id,
            text=new_status
        )

        await self.update_forwarded_post_status(post_info_entity, new_status)
        
        
    async def on_inspect_more_post(self , message):
        self.bot_settings = self.bot_settings_service.get_bot_settings()
        
        data = message.data.decode('utf-8').split("/")
        post_info_entity = self.post_service.get_post_by_id(int(data[1]))

        chat_id = message.chat_id
        message_id = message.message_id
        first_name = message.sender.first_name
        username = message.sender.username or ""

        new_status = MessagesStatus.more_message(first_name, username)
        markup = get_inline_keyboard(post_info_entity["id"])

        # Forward the original message to the inspection group
        fwd_msg_events = await self.client(functions.messages.ForwardMessagesRequest(
            from_peer=post_info_entity["telegramId_ChatId"],
            id=post_info_entity['telegram_mediagroup_postids'],
            to_peer=self.bot_settings["inspectGroupID"],
            top_msg_id=self.bot_settings["inspectGroupTopicID"],
        ))
        
        for updt in fwd_msg_events.updates:
            if type(updt) == UpdateNewChannelMessage and updt.message != '':
                sent_message = updt.message
                break

        # Send a text message with the inspection status
        markup_message = await self.client.send_message(
            entity=self.bot_settings["inspectGroupID"],
            message="وضعیت پست:",
            reply_to=sent_message,
            buttons=get_inspect_more_inline_keyboard(post_info_entity["id"]),
        )

        # Update database with inspected post details
        post_info_telegram_model = {
            "telegramId_ChatId": markup_message.chat_id,
            "telegramId_PostId": markup_message.id,
            "repliyed_TelegramId_ChatId": sent_message.chat_id,
            "repliyed_TelegramId_PostId": sent_message.id,
        }
        self.post_service.inspect_post(post_info_entity, post_info_telegram_model)

        # Edit the original callback query message with new status and markup
        await self.client.edit_message(
            chat_id,
            message=message_id,
            text=new_status,
            buttons=markup,
        )
        
    async def on_license_check_post(self, message):
        self.bot_settings = self.bot_settings_service.get_bot_settings()
        
        data = message.data.decode('utf-8').split("/")
        post_info_entity = self.post_service.get_post_by_id(int(data[1]))

        chat_id = message.chat_id
        message_id = message.message_id
        first_name = message.sender.first_name
        username = message.sender.username or ""

        new_status = MessagesStatus.license_check_message(first_name, username)
        markup = get_inline_keyboard(post_info_entity["id"])

        try:
            
            fwd_msg_events = await self.client(functions.messages.ForwardMessagesRequest(
                from_peer=post_info_entity["telegramId_ChatId"],
                id=post_info_entity['telegram_mediagroup_postids'],
                to_peer=self.bot_settings["licenseGroupID"],
                top_msg_id=self.bot_settings["licenseGroupTopicID"],
            ))
            
            for updt in fwd_msg_events.updates:
                if type(updt) == UpdateNewChannelMessage and updt.message != '':
                    sent_message = updt.message
                    break
        
            # Send a text message with the license check status
            markup_message = await self.client.send_message(
                entity=self.bot_settings["licenseGroupID"],
                message="وضعیت پست:",
                reply_to=sent_message,
                buttons=get_license_check_inline_keyboard(post_info_entity["id"]),
            )

            # Update database with license check post details
            post_info_telegram_model = {
                "telegramId_ChatId": markup_message.chat_id,
                "telegramId_PostId": markup_message.id,
                "repliyed_TelegramId_ChatId": sent_message.chat_id,
                "repliyed_TelegramId_PostId": sent_message.id,
            }
            self.post_service.license_check_post(post_info_entity, post_info_telegram_model)

            # Edit the original callback query message with new status and markup
            await self.client.edit_message(
                chat_id,
                message=message_id,
                text=new_status,
                buttons=markup,
            )

        except Exception as e:
            print(f"An error occurred: {e}")
            
    async def on_set_inspect_topic_group(self , message):
        self.bot_settings = self.bot_settings_service.get_bot_settings()
        
        data = message.data.decode('utf-8').split("/")
        group_id = int(data[1])
        topic_id = int(data[2])

        chat_id = message.chat_id
        message_id = message.message_id

        new_status = "<b>این تاپیک به عنوان تاپیک بازرسی تنظیم شد ✅</b>"

        # Update bot settings
        self.bot_settings["inspectGroupID"] = group_id
        self.bot_settings["inspectGroupTopicID"] = topic_id

        # Save updated bot settings (replace with your actual service call)
        self.bot_settings = self.bot_settings_service.set_bot_settings(self.bot_settings)

        # log_message = f"Set Inspect Topic Group: {group_id} {topic_id}"
        # logger.log_info(log_message)

        # Edit the message with the new status
        await self.client.edit_message(
            chat_id,
            message=message_id,
            text=new_status,
            parse_mode="html"
        )


    async def on_set_license_check_group(self,message):
        self.bot_settings = self.bot_settings_service.get_bot_settings()
        
        data = message.data.decode('utf-8').split("/")
        group_id = int(data[1])
        topic_id = int(data[2])

        chat_id = message.chat_id
        message_id = message.message_id

        new_status = "<b>این تاپیک به عنوان تاپیک بررسی مجوز تنظیم شد ✅</b>"

        # Update bot settings
        self.bot_settings["licenseGroupID"] = group_id
        self.bot_settings["licenseGroupTopicID"] = topic_id

        # Save updated bot settings (replace with your actual service call)
        self.bot_settings = self.bot_settings_service.set_bot_settings(self.bot_settings)

        # log_message = f"Set License Check Topic Group: {group_id} {topic_id}"
        # logger.log_info(log_message)

        # Edit the message with the new status
        await self.client.edit_message(
            chat_id,
            message=message_id,
            text=new_status,
            parse_mode='html',
        )

    async def on_set_main_alternative_group_topic(self,message):
        self.bot_settings = self.bot_settings_service.get_bot_settings()
        
        data = message.data.decode('utf-8').split("/")
        group_id = int(data[1])
        topic_id = int(data[2])

        chat_id = message.chat_id
        message_id = message.message_id

        new_status = "<b>این تاپیک به عنوان تاپیک ارسال پست ها تنظیم شد ✅</b>"

        # Update bot settings
        self.bot_settings["alternativeMainGroupID"] = group_id
        self.bot_settings["alternativeMainGroupTopicID"] = topic_id

        # Save updated bot settings (replace with your actual service call)
        self.bot_settings = self.bot_settings_service.set_bot_settings(self.bot_settings)

        # log_message = f"Set Alternative Topic Group: {group_id} {topic_id}"
        # logger.log_info(log_message)

        # Edit the message with the new status
        await self.client.edit_message(
            chat_id,
            message=message_id,
            text=new_status,
            parse_mode='html',
        )

File: BackEnd/TelegramBot.Bot/CallbackQueryHandler.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using AutoMapper;
using Telegram.Bot;
using Telegram.Bot.Types;
using Telegram.Bot.Types.Enums;
using Telegram.Bot.Types.ReplyMarkups;
using TelegramBot.Application.Interfaces;
using TelegramBot.Application.Services.BotSettings;
using TelegramBot.Application.Services.PostService;
using TelegramBot.Core.Entities;

namespace TelegramBot.Bot
{
    public class CallbackQueryHandler
    {
        IMapper mapper;
        BotSettingsEntity botSettings;
        IBotSettingsService botSettingsService;
        ILoggerService logger;
        IPostService postService;
        private readonly ITelegramBotClient botClient;
        private readonly Update update;
        private readonly CancellationToken cancellationToken;

        public CallbackQueryHandler(ITelegramBotClient botClient, Update update, CancellationToken cancellationToken, BotSettingsEntity botSettings, ILoggerService logger, IMapper mapper, IBotSettingsService botSettingsService, IPostService postService)
        {
            this.botClient = botClient;
            this.update = update;
            this.cancellationToken = cancellationToken;
            this.botSettings = botSettings;
            this.logger = logger;
            this.mapper = mapper;
            this.botSettingsService = botSettingsService;
            this.postService = postService;
        }

        async void UpdateForwardedPostStatus(PostInfoEntity postInfoEntity, string status, InlineKeyboardMarkup? markup)
        {

            var postEvents = postService.GetRemovableEvents(postInfoEntity.Id).ToList();
            foreach (var postEvent in postEvents)
            {
                try
                {
                    await botClient.EditMessageTextAsync(chatId: postEvent.TelegramId_ChatId,
                                            messageId: postEvent.TelegramId_PostId,
                                            text: status,
                                            replyMarkup: markup,
                                            cancellationToken: cancellationToken);
                }
                catch (Exception ex)
                {
                    Console.WriteLine(ex.Message);
                    continue;
                }
            }
        }

        public async void OnSetBridgeChannel(CallbackQuery message, string bridgeChanneID)
        {
            var chatId = message.Message.Chat.Id;
            var messageId = message.Message.MessageId;

            var newStatus = "<b>این کانال به عنوان کانال میانی تنظیم شد ✅</b>";

            var newbotSettings = mapper.Map<BotSettingsModel>(botSettings);
            newbotSettings.BridgeChanneID = long.Parse(bridgeChanneID);

            botSettingsService.SetBotSettings(newbotSettings);

            var logMessage = $"Set Bridge channel: {bridgeChanneID}";
            logger.LogInfo(logMessage);

            await botClient.EditMessageTextAsync(chatId: chatId,
                                                    messageId: messageId,
                                                    text: newStatus,
                                                    parseMode: ParseMode.Html,
                                                    cancellationToken: cancellationToken);
        }

        public async void OnAcceptPost(CallbackQuery message)
        {
            var data = message.Data.Split("/");
            PostInfoEntity postInfoEntity = postService.GetByPostId(int.Parse(data[1]));

            var acceptedPost = postService.AcceptPost(postInfoEntity);

            var chatId = message.Message.Chat.Id;
            var messageId = message.Message.MessageId;
            var FirstName = message.From.FirstName;
            var Username = message.From.Username ?? "";

            var newStatus = MessagesStatus.AcceptMessage(FirstName, Username);

            // Message sentMessage = await botClient.ForwardMessageAsync(
            //             chatId: botSettings.BridgeChanneID,
            //             fromChatId: message.Message.ReplyToMessage.Chat.Id,
            //             messageId: message.Message.ReplyToMessage.MessageId,
            //             cancellationToken: cancellationToken);

            Message acceptMessage = await botClient.SendTextMessageAsync(
                chatId: botSettings.BridgeChanneID,
                text: $"{botSettings.MainChanneID},{acceptedPost.ForwardToChannelDate},{postInfoEntity.TelegramId_PostId},{postInfoEntity.TelegramId_ChatId}",
                cancellationToken: cancellationToken);

            var logMessage = $"Forward to channel: {postInfoEntity.TelegramId_PostId} , {postInfoEntity.TelegramId_ChatId} , postid: {postInfoEntity.Id}";
            logger.LogInfo(logMessage);

            await botClient.EditMessageTextAsync(chatId: chatId,
                                                        messageId: messageId,
                                                        text: newStatus,
                                                        cancellationToken: cancellationToken);

            UpdateForwardedPostStatus(postInfoEntity, newStatus, null);

        }

        public async void OnDeclinePost(CallbackQuery message)
        {
            var data = message.Data.Split("/");
            PostInfoEntity postInfoEntity = postService.GetByPostId(int.Parse(data[1]));

            var chatId = message.Message.Chat.Id;
            var messageId = message.Message.MessageId;
            var FirstName = message.From.FirstName;
            var Username = message.From.Username ?? "";

            postService.DeclinePost(postInfoEntity);

            var newStatus = MessagesStatus.DeclineMessage(FirstName, Username);

            await botClient.EditMessageTextAsync(chatId: chatId,
                                                        messageId: messageId,
                                                        text: newStatus,
                                                        cancellationToken: cancellationToken);

            UpdateForwardedPostStatus(postInfoEntity, newStatus, null);

        }

        public async void OnInspectMorePost(CallbackQuery message)
        {
            var data = message.Data.Split("/");
            PostInfoEntity postInfoEntity = postService.GetByPostId(int.Parse(data[1]));

            var chatId = message.Message.Chat.Id;
            var messageId = message.Message.MessageId;
            var FirstName = message.From.FirstName;
            var Username = message.From.Username ?? "";


            var newStatus = MessagesStatus.MoreMessage(FirstName, Username);
            var markup = Utility.getInlineKeyboard(postInfoEntity.Id);

            Message sentMessage = await botClient.ForwardMessageAsync(
                        chatId: botSettings.InspectGroupID,
                        messageThreadId: botSettings.InspectGroupTopicID,
                        fromChatId: message.Message.ReplyToMessage.Chat.Id,
                        messageId: message.Message.ReplyToMessage.MessageId,
                        cancellationToken: cancellationToken);

            var inlineKeyboard = Utility.getInspectMoreInlineKeyboard(postInfoEntity.Id);

            Message markupMessage = await botClient.SendTextMessageAsync(
                    chatId: botSettings.InspectGroupID,
                    text: "وضعیت پست:",
                    replyMarkup: inlineKeyboard,
                    replyToMessageId: sentMessage.MessageId,
                    cancellationToken: cancellationToken);

            PostInfoTelegramModel postInfoTelegramModel = new()
            {
                TelegramId_ChatId = markupMessage.Chat.Id,
                TelegramId_PostId = markupMessage.MessageId,
                Repliyed_TelegramId_ChatId = sentMessage.Chat.Id,
                Repliyed_TelegramId_PostId = sentMessage.MessageId
            };

            postService.InspectPost(postInfoEntity, postInfoTelegramModel);

            await botClient.EditMessageTextAsync(chatId: chatId,
                                            messageId: messageId,
                                            text: newStatus,
                                            replyMarkup: markup,
                                            cancellationToken: cancellationToken);

        }

        public async void OnLicenseCheckPost(CallbackQuery message)
        {
            var data = message.Data.Split("/");
            PostInfoEntity postInfoEntity = postService.GetByPostId(int.Parse(data[1]));

            var chatId = message.Message.Chat.Id;
            var messageId = message.Message.MessageId;
            var FirstName = message.From.FirstName;
            var Username = message.From.Username ?? "";

            var newStatus = MessagesStatus.LicenseCheckMessage(FirstName, Username);
            var markup = Utility.getInlineKeyboard(postInfoEntity.Id);

            Message sentMessage = await botClient.ForwardMessageAsync(
                        chatId: botSettings.LicenseGroupID,
                        messageThreadId: botSettings.LicenseGroupTopicID,
                        fromChatId: message.Message.ReplyToMessage.Chat.Id,
                        messageId: message.Message.ReplyToMessage.MessageId,
                        cancellationToken: cancellationToken);

            var inlineKeyboard = Utility.getLicenseCheckInlineKeyboard(postInfoEntity.Id);

            Message markupMessage = await botClient.SendTextMessageAsync(
                    chatId: botSettings.LicenseGroupID,
                    text: "وضعیت پست:",
                    replyMarkup: inlineKeyboard,
                    replyToMessageId: sentMessage.MessageId,
                    cancellationToken: cancellationToken);

            PostInfoTelegramModel postInfoTelegramModel = new()
            {
                TelegramId_ChatId = markupMessage.Chat.Id,
                TelegramId_PostId = markupMessage.MessageId,
                Repliyed_TelegramId_ChatId = sentMessage.Chat.Id,
                Repliyed_TelegramId_PostId = sentMessage.MessageId
            };

            postService.LicenseCheckPost(postInfoEntity, postInfoTelegramModel);

            try
            {
                await botClient.EditMessageTextAsync(chatId: chatId,
                                                messageId: messageId,
                                                text: newStatus,
                                                replyMarkup: markup,
                                                cancellationToken: cancellationToken);
            }
            catch (System.Exception)
            {

                return;
            }

        }

        public async void OnSetInspectTopicGroup(CallbackQuery message)
        {
            var data = message.Data.Split("/");
            var groupId = long.Parse(data[1]);
            var topicId = int.Parse(data[2]);

            var chatId = message.Message.Chat.Id;
            var messageId = message.Message.MessageId;

            var newStatus = "<b>این تاپیک به عنوان تاپیک بازرسی تنظیم شد ✅</b>";

            var newbotSettings = mapper.Map<BotSettingsModel>(botSettings);
            newbotSettings.InspectGroupID = groupId;
            newbotSettings.InspectGroupTopicID = topicId;

            botSettingsService.SetBotSettings(newbotSettings);

            var logMessage = $"Set Inspect Topic Group: {groupId} {topicId}";
            logger.LogInfo(logMessage);

            await botClient.EditMessageTextAsync(chatId: chatId,
                                                    messageId: messageId,
                                                    text: newStatus,
                                                    parseMode: ParseMode.Html,
                                                    cancellationToken: cancellationToken);
        }

        public async void OnSetLicenseCheckGroup(CallbackQuery message)
        {
            var data = message.Data.Split("/");
            var groupId = long.Parse(data[1]);
            var topicId = int.Parse(data[2]);

            var chatId = message.Message.Chat.Id;
            var messageId = message.Message.MessageId;

            var newStatus = "<b>این تاپیک به عنوان تاپیک بررسی مجوز تنظیم شد ✅</b>";

            var newbotSettings = mapper.Map<BotSettingsModel>(botSettings);
            newbotSettings.LicenseGroupID = groupId;
            newbotSettings.LicenseGroupTopicID = topicId;

            botSettingsService.SetBotSettings(newbotSettings);

            var logMessage = $"Set License Check Topic Group: {groupId} {topicId}";
            logger.LogInfo(logMessage);

            await botClient.EditMessageTextAsync(chatId: chatId,
                                                    messageId: messageId,
                                                    text: newStatus,
                                                    parseMode: ParseMode.Html,
                                                    cancellationToken: cancellationToken);
        }

        public async void OnSetMainAlternativeGroupTopic(CallbackQuery message)
        {
            var data = message.Data.Split("/");
            var groupId = long.Parse(data[1]);
            var topicId = int.Parse(data[2]);

            var chatId = message.Message.Chat.Id;
            var messageId = message.Message.MessageId;

            var newStatus = "<b>این تاپیک به عنوان تاپیک ارسال پست ها تنظیم شد ✅</b>";

            var newbotSettings = mapper.Map<BotSettingsModel>(botSettings);
            newbotSettings.AlternativeMainGroupID = groupId;
            newbotSettings.AlternativeMainGroupTopicID = topicId;

            botSettingsService.SetBotSettings(newbotSettings);

            var logMessage = $"Set Alternative Topic Group: {groupId} {topicId}";
            logger.LogInfo(logMessage);

            await botClient.EditMessageTextAsync(chatId: chatId,
                                                    messageId: messageId,
                                                    text: newStatus,
                                                    parseMode: ParseMode.Html,
                                                    cancellationToken: cancellationToken);
        }

    }
}